File: app/src/main/java/be/digitalia/fosdem/model/Event.java
package be.digitalia.fosdem.model;

import android.os.Parcel;
import android.os.Parcelable;
import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Embedded;
import androidx.room.TypeConverters;
import be.digitalia.fosdem.api.FosdemUrls;
import be.digitalia.fosdem.db.converters.NullableDateTypeConverters;
import be.digitalia.fosdem.utils.DateUtils;

import java.util.Date;

public class Event implements Parcelable {

	private long id;
	@Embedded(prefix = "day_")
	@NonNull
	private Day day;
	@ColumnInfo(name = "start_time")
	@TypeConverters({NullableDateTypeConverters.class})
	private Date startTime;
	@ColumnInfo(name = "end_time")
	@TypeConverters({NullableDateTypeConverters.class})
	private Date endTime;
	@ColumnInfo(name = "room_name")
	private String roomName;
	private String slug;
	private String title;
	@ColumnInfo(name = "subtitle")
	private String subTitle;
	@Embedded(prefix = "track_")
	@NonNull
	private Track track;
	@ColumnInfo(name = "abstract")
	private String abstractText;
	private String description;
	@ColumnInfo(name = "persons")
	private String personsSummary;

	public Event() {
	}

	public long getId() {
		return id;
	}

	public void setId(long id) {
		this.id = id;
	}

	@NonNull
	public Day getDay() {
		return day;
	}

	public void setDay(@NonNull Day day) {
		this.day = day;
	}

	public Date getStartTime() {
		return startTime;
	}

	public void setStartTime(Date startTime) {
		this.startTime = startTime;
	}

	public Date getEndTime() {
		return endTime;
	}

	public void setEndTime(Date endTime) {
		this.endTime = endTime;
	}

	public boolean isRunningAtTime(long time) {
		return (startTime != null) && (endTime != null) && (startTime.getTime() < time) && (time < endTime.getTime());
	}

	/**
	 * @return The event duration in minutes
	 */
	public int getDuration() {
		if ((startTime == null) || (endTime == null)) {
			return 0;
		}
		return (int) ((this.endTime.getTime() - this.startTime.getTime()) / android.text.format.DateUtils.MINUTE_IN_MILLIS);
	}

	public String getRoomName() {
		return (roomName == null) ? "" : roomName;
	}

	public void setRoomName(String roomName) {
		this.roomName = roomName;
	}

	public String getSlug() {
		return slug;
	}

	public void setSlug(String slug) {
		this.slug = slug;
	}

	public String getUrl() {
		return FosdemUrls.getEvent(slug, DateUtils.getYear(getDay().getDate().getTime()));
	}

	public String getTitle() {
		return title;
	}

	public void setTitle(String title) {
		this.title = title;
	}

	public String getSubTitle() {
		return subTitle;
	}

	public void setSubTitle(String subTitle) {
		this.subTitle = subTitle;
	}

	@NonNull
	public Track getTrack() {
		return track;
	}

	public void setTrack(@NonNull Track track) {
		this.track = track;
	}

	public String getAbstractText() {
		return abstractText;
	}

	public void setAbstractText(String abstractText) {
		this.abstractText = abstractText;
	}

	public String getDescription() {
		return description;
	}

	public void setDescription(String description) {
		this.description = description;
	}

	@NonNull
	public String getPersonsSummary() {
		if (personsSummary != null) {
			return personsSummary;
		}
		return "";
	}

	public void setPersonsSummary(String personsSummary) {
		this.personsSummary = personsSummary;
	}

	@NonNull
	@Override
	public String toString() {
		return title;
	}

	@Override
	public int hashCode() {
		return (int) (id ^ (id >>> 32));
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		Event other = (Event) obj;
		return id == other.id;
	}

	@Override
	public int describeContents() {
		return 0;
	}

	@Override
	public void writeToParcel(Parcel out, int flags) {
		out.writeLong(id);
		day.writeToParcel(out, flags);
		out.writeLong((startTime == null) ? 0L : startTime.getTime());
		out.writeLong((endTime == null) ? 0L : endTime.getTime());
		out.writeString(roomName);
		out.writeString(slug);
		out.writeString(title);
		out.writeString(subTitle);
		track.writeToParcel(out, flags);
		out.writeString(abstractText);
		out.writeString(description);
		out.writeString(personsSummary);
	}

	public static final Parcelable.Creator<Event> CREATOR = new Parcelable.Creator<Event>() {
		public Event createFromParcel(Parcel in) {
			return new Event(in);
		}

		public Event[] newArray(int size) {
			return new Event[size];
		}
	};

	Event(Parcel in) {
		id = in.readLong();
		day = Day.CREATOR.createFromParcel(in);
		long time = in.readLong();
		if (time != 0L) {
			startTime = new Date(time);
		}
		time = in.readLong();
		if (time != 0L) {
			endTime = new Date(time);
		}
		roomName = in.readString();
		slug = in.readString();
		title = in.readString();
		subTitle = in.readString();
		track = Track.CREATOR.createFromParcel(in);
		abstractText = in.readString();
		description = in.readString();
		personsSummary = in.readString();
	}
}

File: app/src/main/java/be/digitalia/fosdem/model/Event.kt
package be.digitalia.fosdem.model

import android.os.Parcelable
import androidx.room3.ColumnInfo
import androidx.room3.ColumnTypeConverters
import androidx.room3.DatabaseView
import androidx.room3.Embedded
import be.digitalia.fosdem.db.converters.NullableInstantTypeConverters
import be.digitalia.fosdem.db.converters.NullableZoneOffsetTypeConverters
import be.digitalia.fosdem.utils.InstantParceler
import be.digitalia.fosdem.utils.ZoneOffsetParceler
import kotlinx.parcelize.Parcelize
import kotlinx.parcelize.WriteWith
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import kotlin.time.Duration
import kotlin.time.Instant
import kotlin.time.toJavaInstant

@DatabaseView(
    value = """SELECT e.id, e.start_time, e.start_time_offset, e.end_time, e.room_name, e.url,
        et.title, et.subtitle, e.abstract, e.description, e.feedback_url, GROUP_CONCAT(p.name, ', ') AS persons,
        e.day_index, d.date AS day_date, d.start_time AS day_start_time, d.end_time AS day_end_time,
        e.track_id, t.name AS track_name, t.type AS track_type
        FROM events e
        JOIN events_titles et ON e.id = et.`rowid`
        JOIN days d ON e.day_index = d.`index`
        JOIN tracks t ON e.track_id = t.id
        LEFT JOIN events_persons ep ON e.id = ep.event_id
        LEFT JOIN persons p ON ep.person_id = p.`rowid`
        GROUP BY e.id""",
    viewName = "events_view"
)
@Parcelize
data class Event(
        val id: Long,
        @Embedded(prefix = "day_")
        val day: Day,
        @ColumnInfo(name = "start_time")
        @field:ColumnTypeConverters(NullableInstantTypeConverters::class)
        val startTime: @WriteWith<InstantParceler> Instant? = null,
        @ColumnInfo(name = "start_time_offset")
        @field:ColumnTypeConverters(NullableZoneOffsetTypeConverters::class)
        val startTimeOffset: @WriteWith<ZoneOffsetParceler> ZoneOffset? = null,
        @ColumnInfo(name = "end_time")
        @field:ColumnTypeConverters(NullableInstantTypeConverters::class)
        val endTime: @WriteWith<InstantParceler> Instant? = null,
        @ColumnInfo(name = "room_name")
        val roomName: String?,
        val url: String?,
        val title: String?,
        @ColumnInfo(name = "subtitle")
        val subTitle: String?,
        @Embedded(prefix = "track_")
        val track: Track,
        @ColumnInfo(name = "abstract")
        val abstractText: String?,
        val description: String?,
        @ColumnInfo(name = "feedback_url")
        val feedbackUrl: String?,
        @ColumnInfo(name = "persons")
        val personsSummary: String?
) : Parcelable {

    fun startTime(zoneOverride: ZoneId?): LocalDateTime? {
        val zone = zoneOverride ?: startTimeOffset
        return if (startTime != null && zone != null)
            LocalDateTime.ofInstant(startTime.toJavaInstant(), zone)
        else null
    }

    fun endTime(zoneOverride: ZoneId?): LocalDateTime? {
        val zone = zoneOverride ?: startTimeOffset
        return if (endTime != null && zone != null)
            LocalDateTime.ofInstant(endTime.toJavaInstant(), zone)
        else null
    }

    fun isRunningAtTime(time: Instant): Boolean {
        return startTime != null && endTime != null && time >= startTime && time < endTime
    }

    val duration: Duration
        get() = if (startTime == null || endTime == null) {
            Duration.ZERO
        } else {
            endTime - startTime
        }

    override fun toString(): String = title.orEmpty()
}